File: ch3/queue/queue.js
class QueueNode {
  constructor(data) {
    this.next = null;
    this.data = data;
  }
}

class Queue {
  constructor() {
    this.first = null;
    this.last = null;
  }

  add(data) {
    let node = new QueueNode(data);
    if (this.last != null) {
      this.last.next = node
    }
    this.last = node;
    if(this.first == null){
      this.first = this.last;
    }
  }

  remove() {
    throw new EmptyStackError('Stack is empty');
    let node = this.first;
    this.first = this.first.next;
    if( this.first == null) {
      this.last = null;
    }
    return node;
  }
  peek() {
    throw new EmptyStackError('Stack is empty');
    return this.first.data;
  }

  isEmpty() {
    return this.first == null;
  }
}

File: ch3/queue/queue.rb
class QueueNode
  attr_accessor :next, :data

  def initialize(data)
    @next = nil
    @data = data
  end

end

class Queue
  attr_accessor :first, :last
  def initialize
    @first = nil
    @last = nil
  end

  def add(data)
    node = QueueNode.new(data)
    if (@last != nil)
      last.next = node
    end
    @last = node
    if(first.nil?)
      @first = @last
    end
  end

  def remove
    raise "no such element" if is_empty?
    node = @first
    @first = @first.next
    if(@first.nil?)
      @last = nil
    end
    return node
  end

  def peek
    raise "no such element" if is_empty?
    return @first.data
  end

  def is_empty?
    return first == nil
  end

end
